File: src/ball/Ball.ts
import { nanoid } from 'nanoid';
import type { Ball as BallType, BallVelocity, Position } from '../types/index.js';
import { BallState } from '../types/index.js';

/**
 * Create a new ball at a spawn point
 */
export function createBall(
  spawnPointId: string,
  position: Position,
  alliance: 'red' | 'blue' | null = null
): BallType {
  return {
    id: nanoid(),
    state: BallState.ON_FIELD,
    position: { ...position },
    height: 0,
    velocity: { vx: 0, vy: 0, vz: 0 },
    heldByRobotId: null,
    shotByRobotId: null,
    shotFromPosition: null,
    targetPosition: null,
    spawnPointId,
    alliance,
    lastUpdateTick: 0,
    claimedByRobotId: null,
    claimedAtTick: null,
  };
}

/**
 * Ball entity class for managing ball state
 */
export class Ball {
  data: BallType;

  constructor(data: BallType) {
    this.data = data;
  }

  get id(): string {
    return this.data.id;
  }

  get state(): BallState {
    return this.data.state;
  }

  get position(): Position {
    return this.data.position;
  }

  get height(): number {
    return this.data.height;
  }

  get velocity(): BallVelocity {
    return this.data.velocity;
  }

  /**
   * Check if ball is available for pickup
   */
  isAvailable(): boolean {
    return this.data.state === BallState.ON_FIELD;
  }

  /**
   * Check if ball is in flight
   */
  isInFlight(): boolean {
    return this.data.state === BallState.IN_FLIGHT;
  }

  /**
   * Check if ball is held
   */
  isHeld(): boolean {
    return this.data.state === BallState.HELD;
  }

  /**
   * Check if ball is claimed by a robot
   */
  isClaimed(): boolean {
    return this.data.claimedByRobotId !== null;
  }

  /**
   * Check if ball is claimed by a specific robot
   */
  isClaimedBy(robotId: string): boolean {
    return this.data.claimedByRobotId === robotId;
  }

  /**
   * Get the robot ID that claimed this ball
   */
  getClaimingRobotId(): string | null {
    return this.data.claimedByRobotId;
  }

  /**
   * Claim this ball for pickup (alliance coordination)
   * Returns true if claim was successful
   */
  claim(robotId: string, tick: number): boolean {
    // Can't claim if already claimed by someone else
    if (this.data.claimedByRobotId !== null && this.data.claimedByRobotId !== robotId) {
      return false;
    }
    // Can't claim if not available
    if (!this.isAvailable()) {
      return false;
    }
    this.data.claimedByRobotId = robotId;
    this.data.claimedAtTick = tick;
    return true;
  }

  /**
   * Release claim on this ball
   */
  releaseClaim(robotId?: string): void {
    // If robotId specified, only release if claimed by that robot
    if (robotId && this.data.claimedByRobotId !== robotId) {
      return;
    }
    this.data.claimedByRobotId = null;
    this.data.claimedAtTick = null;
  }

  /**
   * Check if claim has expired (claims expire after 3 seconds / 180 ticks)
   */
  isClaimExpired(currentTick: number, expirationTicks: number = 180): boolean {
    if (this.data.claimedAtTick === null) return true;
    return currentTick - this.data.claimedAtTick > expirationTicks;
  }

  /**
   * Pick up the ball (robot takes possession)
   */
  pickup(robotId: string, tick: number): void {
    this.data.state = BallState.HELD;
    this.data.heldByRobotId = robotId;
    this.data.velocity = { vx: 0, vy: 0, vz: 0 };
    this.data.lastUpdateTick = tick;
    // Clear any claim when ball is picked up
    this.data.claimedByRobotId = null;
    this.data.claimedAtTick = null;
  }

  /**
   * Shoot the ball towards a target
   */
  shoot(
    robotId: string,
    fromPosition: Position,
    targetPosition: Position,
    initialVelocity: BallVelocity,
    tick: number
  ): void {
    this.data.state = BallState.IN_FLIGHT;
    this.data.position = { ...fromPosition };
    this.data.heldByRobotId = null;
    this.data.shotByRobotId = robotId;
    this.data.shotFromPosition = { ...fromPosition };
    this.data.targetPosition = { ...targetPosition };
    this.data.velocity = { ...initialVelocity };
    this.data.lastUpdateTick = tick;
  }

  /**
   * Ball lands on ground (or ramp surface)
   */
  land(position: Position, tick: number, height: number = 0): void {
    this.data.state = BallState.ON_FIELD;
    this.data.position = { ...position };
    this.data.height = Math.max(0, height);
    this.data.shotByRobotId = null;
    this.data.shotFromPosition = null;
    this.data.targetPosition = null;
    this.data.lastUpdateTick = tick;
  }

  /**
   * Ball is scored
   */
  score(tick: number): void {
    this.data.state = BallState.SCORED;
    this.data.velocity = { vx: 0, vy: 0, vz: 0 };
    this.data.lastUpdateTick = tick;
  }

  /**
   * Ball goes out of bounds
   */
  outOfBounds(tick: number): void {
    this.data.state = BallState.OUT_OF_BOUNDS;
    this.data.velocity = { vx: 0, vy: 0, vz: 0 };
    this.data.lastUpdateTick = tick;
  }

  /**
   * Respawn ball at its spawn point
   */
  respawn(position: Position, tick: number): void {
    this.data.state = BallState.ON_FIELD;
    this.data.position = { ...position };
    this.data.height = 0;
    this.data.velocity = { vx: 0, vy: 0, vz: 0 };
    this.data.heldByRobotId = null;
    this.data.shotByRobotId = null;
    this.data.shotFromPosition = null;
    this.data.targetPosition = null;
    this.data.lastUpdateTick = tick;
    this.data.claimedByRobotId = null;
    this.data.claimedAtTick = null;
  }

  /**
   * Respawn ball with specific position, height, and velocity (for scoring zone exits)
   */
  respawnWithVelocity(
    position: Position,
    height: number,
    velocity: BallVelocity,
    tick: number
  ): void {
    this.data.state = BallState.IN_FLIGHT;
    this.data.position = { ...position };
    this.data.height = height;
    this.data.velocity = { ...velocity };
    this.data.heldByRobotId = null;
    this.data.shotByRobotId = null;
    this.data.shotFromPosition = null;
    this.data.targetPosition = null;
    this.data.lastUpdateTick = tick;
    this.data.claimedByRobotId = null;
    this.data.claimedAtTick = null;
  }

  /**
   * Update ball position
   */
  setPosition(position: Position): void {
    this.data.position = { ...position };
  }

  /**
   * Update ball height
   */
  setHeight(height: number): void {
    this.data.height = Math.max(0, height);
  }

  /**
   * Update ball velocity
   */
  setVelocity(velocity: BallVelocity): void {
    this.data.velocity = { ...velocity };
  }

  /**
   * Clone ball data
   */
  cloneData(): BallType {
    return {
      ...this.data,
      position: { ...this.data.position },
      velocity: { ...this.data.velocity },
      shotFromPosition: this.data.shotFromPosition
        ? { ...this.data.shotFromPosition }
        : null,
      targetPosition: this.data.targetPosition
        ? { ...this.data.targetPosition }
        : null,
    };
  }
}

File: src/frc_sim/entities/ball.py
"""Ball entity class."""

from __future__ import annotations
from typing import Optional
from dataclasses import dataclass, field

from ..types.enums import BallState
from ..types.schemas import Position, BallVelocity, BallData


@dataclass
class Ball:
    """Ball entity with state and behavior."""

    id: str
    state: BallState
    position: Position
    height: float
    velocity: BallVelocity
    spawn_point_id: str
    last_update_tick: int
    held_by_robot_id: Optional[str] = None
    shot_by_robot_id: Optional[str] = None
    shot_from_position: Optional[Position] = None
    target_position: Optional[Position] = None
    alliance: Optional[str] = None
    claimed_by_robot_id: Optional[str] = None
    claimed_at_tick: Optional[int] = None

    def clone_data(self) -> BallData:
        """Clone the ball data for serialization."""
        return BallData(
            id=self.id,
            state=self.state.value,
            position=self.position,
            height=self.height,
            velocity=self.velocity,
            spawnPointId=self.spawn_point_id,
            lastUpdateTick=self.last_update_tick,
            heldByRobotId=self.held_by_robot_id,
            shotByRobotId=self.shot_by_robot_id,
            shotFromPosition=self.shot_from_position,
            targetPosition=self.target_position,
            alliance=self.alliance,
            claimedByRobotId=self.claimed_by_robot_id,
            claimedAtTick=self.claimed_at_tick,
        )

    def is_available(self) -> bool:
        """Check if ball is available for pickup."""
        return self.state == BallState.ON_FIELD and self.claimed_by_robot_id is None

    def is_held(self) -> bool:
        """Check if ball is held by a robot."""
        return self.state == BallState.HELD

    def is_in_flight(self) -> bool:
        """Check if ball is in flight."""
        return self.state == BallState.IN_FLIGHT

    def is_scored(self) -> bool:
        """Check if ball has been scored."""
        return self.state == BallState.SCORED

    def is_claimed(self) -> bool:
        """Check if ball has been claimed for pickup."""
        return self.claimed_by_robot_id is not None

    def is_claim_expired(self, current_tick: int, expiration_ticks: int = 180) -> bool:
        """Check if the claim has expired (default 3 seconds at 60 ticks/sec)."""
        if self.claimed_at_tick is None:
            return True
        return current_tick - self.claimed_at_tick > expiration_ticks

    def get_claiming_robot_id(self) -> Optional[str]:
        """Get the ID of the robot claiming this ball."""
        return self.claimed_by_robot_id

    def set_position(self, pos: Position) -> None:
        """Set ball position."""
        self.position = pos

    def set_height(self, height: float) -> None:
        """Set ball height."""
        self.height = height

    def set_velocity(self, vel: BallVelocity) -> None:
        """Set ball velocity."""
        self.velocity = vel

    def claim(self, robot_id: str, tick: int) -> None:
        """Claim the ball for pickup."""
        self.claimed_by_robot_id = robot_id
        self.claimed_at_tick = tick

    def release_claim(self) -> None:
        """Release the claim on the ball."""
        self.claimed_by_robot_id = None
        self.claimed_at_tick = None

    def pickup(self, robot_id: str, tick: int) -> None:
        """Mark ball as picked up by a robot."""
        self.state = BallState.HELD
        self.held_by_robot_id = robot_id
        self.last_update_tick = tick
        self.release_claim()

    def shoot(
        self,
        robot_id: str,
        from_position: Position,
        target_position: Position,
        velocity: BallVelocity,
        tick: int,
    ) -> None:
        """Mark ball as shot."""
        self.state = BallState.IN_FLIGHT
        self.held_by_robot_id = None
        self.shot_by_robot_id = robot_id
        self.shot_from_position = from_position
        self.target_position = target_position
        self.velocity = velocity
        self.last_update_tick = tick

    def land(self, position: Position, tick: int, height: float = 0.0) -> None:
        """Land the ball on the field."""
        self.state = BallState.ON_FIELD
        self.position = position
        self.height = height
        self.shot_by_robot_id = None
        self.shot_from_position = None
        self.target_position = None
        self.last_update_tick = tick

    def score(self, tick: int) -> None:
        """Mark ball as scored."""
        self.state = BallState.SCORED
        self.last_update_tick = tick

    def respawn(self, position: Position, tick: int) -> None:
        """Respawn the ball at its spawn point."""
        self.state = BallState.ON_FIELD
        self.position = position
        self.height = 0.0
        self.velocity = BallVelocity(vx=0.0, vy=0.0, vz=0.0)
        self.held_by_robot_id = None
        self.shot_by_robot_id = None
        self.shot_from_position = None
        self.target_position = None
        self.last_update_tick = tick
        self.release_claim()
